docs/source/notes/randomness.rst

```diff
@@ -2,75 +2,115 @@ Reproducibility
+    save time in development by facilitating experimentation, debugging, and
+    regression testing.
+
+Controlling sources of randomness
+.................................
 
-PyTorch
-.......
+PyTorch random number generator
+-------------------------------
 You can use :meth:`torch.manual_seed()` to seed the RNG for all devices (both
 CPU and CUDA)::
 
     import torch
     torch.manual_seed(0)
 
+Random number generators in other libraries
+-------------------------------------------
+If you or any of the libraries you are using rely on NumPy, you can seed the global
+NumPy RNG with::
 
-There are some PyTorch functions that use CUDA functions that can be a source
-of nondeterminism. One class of such CUDA functions are atomic operations,
-in particular :attr:`atomicAdd`, which can lead to the order of additions being
-nondetermnistic. Because floating-point addition is not perfectly associative
-for floating-point operands, :attr:`atomicAdd` with floating-point operands can
-introduce different floating-point rounding errors on each evaluation, which
-introduces a source of nondeterministic variance (aka noise) in the result.
-
-PyTorch functions that use :attr:`atomicAdd` in the forward kernels include
-:meth:`torch.Tensor.index_add_`, :meth:`torch.Tensor.scatter_add_`,
-:meth:`torch.bincount`.
-
-A number of operations have backwards kernels that use :attr:`atomicAdd`,
-including :meth:`torch.nn.functional.embedding_bag`,
-:meth:`torch.nn.functional.ctc_loss`, :meth:`torch.nn.functional.interpolate`,
-and many forms of pooling, padding, and sampling.
-
-There is currently no simple way of avoiding nondeterminism in these functions.
-
-Additionally, the backward path for :meth:`repeat_interleave` operates
-nondeterministically on the CUDA backend because :meth:`repeat_interleave`
-is implemented using :meth:`index_select`, the backward path for
-which is implemented using :meth:`index_add_`, which is known to operate
-nondeterministically (in the forward direction) on the CUDA backend (see above).
-
-cuDNN
-.....
-When running on the cuDNN backend, two further options must be set::
-
-    torch.backends.cudnn.deterministic = True
-    torch.backends.cudnn.benchmark = False
+    import numpy as np
+    np.random.seed(0)
 
-On some versions of cuDNN and CUDA, RNN and LSTM may have non-deterministic behavior.
+However, some applications and libraries may use NumPy Random Generator objects,
+not the global RNG
+(`<https://numpy.org/doc/stable/reference/random/generator.html>`_), and those will
+need to be seeded consistently as well.
+
+If you are using any other libraries that use random number generators, refer to
+the documentation for those libraries to see how to set consistent seeds for them.
+
+CUDA convolution benchmarking
+-----------------------------
+CUDA convolution operations have a performance enhancement feature that can introduce
+randomness between multiple executions of an application. You can disable this feature
```

Comment:
I would like to suggest a slight tweak to this paragraph and the next one: what if `torch.backends.cudnn.benchmark` was introduced first? The flow could be something like this:

- the cuDNN library for CUDA can be a source of nondeterminism across executions
- when calling a convolution using cuDNN, it has a feature where it will benchmark multiple algorithms to see which is fastest; once selected, that algorithm is consistently used during that process
- due to benchmarking noise and different hardware, the benchmarking may select different algorithms on subsequent runs, even on the same machine
- disabling this benchmarking feature with `torch.backends.cudnn.benchmark = False` will cause cuDNN to deterministically select an algorithm, possibly at the cost of reduced performance